File: Client/BetterMusicPlayer/Behavior/Types/PlaybackState.swift
//
//  PlaybackState.swift
//  BetterMusicPlayer
//
//  Created by John Butler on 3/28/26.
//

import Foundation

struct PlaybackState {
    let device: Device
    let repeatState: RepeatState
    let shuffleState: Bool
    let context: Context?
    let lastChange: Date
    let progress: TimeInterval?
    let isPlaying: Bool
    let item: PlayableItem?
    let allowedActions: AllowedActionSet
    
    
    init?(json: Dictionary<String, AnyObject>){
        guard let deviceJson = json["device"] as? Dictionary<String, AnyObject>,
              let device = Device(json: deviceJson),
              let repeateStateString = json["repeat"] as? String,
              let repeatState = RepeatState(string: repeateStateString),
              let shuffleState = json["shuffle"] as? Bool,
              let contextJson = json["context"] as? Dictionary<String, AnyObject>?,
              let lastChangeMSInst = json["timestamp"] as? Double,
              let progressMS = json["progress_ms"] as? Double?,
              let isPlaying = json["is_playing"] as? Bool,
              let itemJson = json["item"] as? Dictionary<String, AnyObject>?,
              let itemType = json["currently_playing_type"] as? String,
              let actionsJson = json["actions"] as? Dictionary<String, AnyObject>,
              let actions = AllowedActionSet(json: actionsJson)
        else {return nil;}
        
        self.device = device;
        self.repeatState = repeatState;
        self.shuffleState = shuffleState;
        if contextJson == nil {
            self.context = nil;
        } else {
            guard let context = Context(json: contextJson!) else {return nil;}
            self.context = context;
        }
        self.lastChange = Date(timeIntervalSince1970: lastChangeMSInst / 1000)
        if progressMS == nil {
            self.progress = nil;
        } else {
            self.progress = TimeInterval(progressMS! / 1000)
        }
        self.isPlaying = isPlaying
        if let itemJson = itemJson {
            switch itemType {
            case "track":
                guard let track = Track(json: itemJson) else {return nil;}
                self.item = .TrackItem(track)
            case "episode":
                guard let episode = Episode(json: itemJson) else {return nil;}
                self.item = .EpisodeItem(episode)
            case "ad": self.item = nil;
            case "unknown": self.item = nil;
            default: return nil;
            }
        } else {
            self.item = nil;
        }
        self.allowedActions = actions;
    }
    
    
    var uri: SpotifyURI? {
        switch item {
        case .TrackItem(let track): return track.uri;
        case .EpisodeItem(let episode): return episode.uri;
        case nil: return nil;
        }
    }
}

enum RepeatState {
    case Off
    case Track
    case Context
    
    init?(string: String){
        switch string {
        case "off": self = .Off;
        case "track": self = .Track;
        case "context": self = .Context;
        default: return nil;
        }
    }
}

struct AllowedActionSet {
    let interruptPlayback: Bool?
    let pause: Bool?
    let resume: Bool?
    let seek: Bool?
    let skipNext: Bool?
    let skipPrevious: Bool?
    let toggleRepeatContext: Bool?
    let toggleShuffle: Bool?
    let toggleRepeatTrack: Bool?
    let transferPlayback: Bool?
    
    init?(json: Dictionary<String, AnyObject>){
        guard let interruptPlayback = json["interrupting_playback"] as? Bool?,
              let pause = json["pausing"] as? Bool?,
              let resume = json["resuming"] as? Bool?,
              let seek = json["seeking"] as? Bool?,
              let skipNext = json["skipping_next"] as? Bool?,
              let skipPrevious = json["skipping_prev"] as? Bool?,
              let toggleRepeatContext = json["toggling_repeat_context"] as? Bool?,
              let toggleShuffle = json["toggling_shuffle"] as? Bool?,
              let toggleRepeatTrack = json["toggling_repeat_track"] as? Bool?,
              let transferPlayback = json["transfer_playback"] as? Bool?
        else {return nil;}
        
        self.interruptPlayback = interruptPlayback;
        self.pause = pause;
        self.resume = resume;
        self.seek = seek;
        self.skipNext = skipNext;
        self.skipPrevious = skipPrevious;
        self.toggleRepeatContext = toggleRepeatContext;
        self.toggleShuffle = toggleShuffle;
        self.toggleRepeatTrack = toggleRepeatTrack;
        self.transferPlayback = transferPlayback;
    }
}

File: Server/src/core_types/playback_state.rs
use std::time::{Duration, SystemTime};

use crate::core_types::{SpotifyURI, TrackID};

use super::{context::Context, device::Device, episode::Episode, track::Track};
use serde::de::Error;

#[derive(Debug, Clone)]
pub struct PlaybackState {
    //device: Device,
    //repeat_state: RepeatState,
    //is_shuffling: bool,
    context: Option<Context>,
    last_change_inst: SystemTime,
    progress: Option<Duration>,
    item: PlayingItem,
    allowed_actions: ActionSet,
}

impl PlaybackState {
    pub fn is_playing(&self) -> bool {
        self.item != PlayingItem::Nothing
    }

    pub fn is_playing_track(&self, track_id: &TrackID) -> bool {
        matches!(&self.item, PlayingItem::Track(track) if track == track_id)
    }

    pub fn uri(&self) -> Option<SpotifyURI> {
        match &self.item {
            PlayingItem::Track(track) => Some(track.uri()),
            PlayingItem::Episode(episode) => Some(episode.uri()),
            _ => None,
        }
    }

    pub fn time_remaining(&self) -> Option<Duration> {
        self.progress.and_then(|progress| {
            if let PlayingItem::Track(track) = &self.item {
                Some(*track.duration() - progress)
            } else {
                None
            }
        })
    }
}

impl<'de> serde::Deserialize<'de> for PlaybackState {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        RawPlaybackState::deserialize(deserializer)?
            .try_into()
            .map_err(|_| D::Error::custom("invalid inner playback state data"))
    }
}

#[derive(Debug, Clone, serde::Deserialize)]
struct RawPlaybackState {
    //device: Device,
    //repeat_state: RepeatState,
    //shuffle_state: bool,
    context: Option<Context>,
    timestamp: u64,
    progress_ms: Option<u64>,
    is_playing: bool,
    item: Option<Track>,
    currently_playing_type: String,
    actions: RawActionSetWrapper,
}

impl TryInto<PlaybackState> for RawPlaybackState {
    type Error = ();

    fn try_into(self) -> Result<PlaybackState, Self::Error> {
        let item: PlayingItem = if self.is_playing {
            match self.currently_playing_type.as_str() {
                "track" => PlayingItem::Track(self.item.unwrap()),
                "episode" => panic!("received an episode. Not handled"),
                "ad" => PlayingItem::Ad,
                _ => PlayingItem::Unknown,
            }
        } else {
            PlayingItem::Nothing
        };

        Ok(PlaybackState {
            //device: self.device,
            //repeat_state: self.repeat_state,
            //is_shuffling: self.shuffle_state,
            context: self.context,
            last_change_inst: SystemTime::UNIX_EPOCH + Duration::from_millis(self.timestamp),
            progress: self.progress_ms.map(Duration::from_millis),
            item,
            allowed_actions: self.actions.disallows,
        })
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PlayingItem {
    Track(Track),
    Episode(Episode),
    Ad,
    Unknown,
    Nothing,
}

#[derive(Debug, Clone, Copy, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
pub enum RepeatState {
    #[serde(rename = "off")]
    Off,
    #[serde(rename = "track")]
    Track,
    #[serde(rename = "context")]
    Context,
}

#[derive(Debug, Clone, PartialEq, serde::Deserialize, Eq)]
pub struct ActionSet {
    interrupting_playback: Option<bool>,
    pausing: Option<bool>,
    resuming: Option<bool>,
    seeking: Option<bool>,
    skipping_next: Option<bool>,
    skipping_previous: Option<bool>,
    toggling_repeat_context: Option<bool>,
    toggling_shuffle: Option<bool>,
    toggling_repeat_track: Option<bool>,
    transferring_playback: Option<bool>,
}

#[derive(Debug, Clone, serde::Deserialize)]
struct RawActionSetWrapper {
    disallows: ActionSet,
}

impl From<RawActionSetWrapper> for ActionSet {
    fn from(from: RawActionSetWrapper) -> Self {
        ActionSet {
            interrupting_playback: from.disallows.interrupting_playback.map(|a| !a),
            pausing: from.disallows.pausing.map(|a| !a),
            resuming: from.disallows.resuming.map(|a| !a),
            seeking: from.disallows.seeking.map(|a| !a),
            skipping_next: from.disallows.skipping_next.map(|a| !a),
            skipping_previous: from.disallows.skipping_previous.map(|a| !a),
            toggling_repeat_context: from.disallows.toggling_repeat_context.map(|a| !a),
            toggling_shuffle: from.disallows.toggling_shuffle.map(|a| !a),
            toggling_repeat_track: from.disallows.toggling_repeat_track.map(|a| !a),
            transferring_playback: from.disallows.transferring_playback.map(|a| !a),
        }
    }
}
